File: matlab/tools/filters/notch_filter.m
function values = notch_filter(values,fs,varargin)
% Apply a notch filter to iEEG data to remove interference at a specified frequency.
%
% Parameters:
% - values (numeric array): Matrix representing EEG data. Each column is a channel, and each row is a time point.
% - fs (numeric): Sampling frequency of the EEG data.
% - notch_freq (numeric, optional): Frequency to notch filter (default is 60 Hz).
% - order (numeric, optional): Order of the notch filter (default is 4).
%
% Returns:
% - values (numeric array): Matrix with notch-filtered EEG data.
%
% Example:
% values = randn(100, 5);  % Replace with your actual iEEG data
% fs = 250;
% notch_filtered_values = notch_filter(values, fs);
% notch_filtered_values = notch_filter(values, fs, [], 8);
% notch_filtered_values = notch_filter(values, fs, 120);

defaults = {60,4};

for i = 1:length(varargin)
    if isempty(varargin{i})
        varargin{i} = defaults{i};
    end
end

p = inputParser;
addRequired(p, 'values', @isnumeric);
addRequired(p, 'fs', @isnumeric);
addOptional(p, 'notch_freq', defaults{1}, @isnumeric);
addOptional(p, 'order', defaults{2}, @isinteger);

parse(p, values, fs, varargin{:});

values = p.Results.values;
fs = p.Results.fs;
notch_freq = p.Results.notch_freq;
order = p.Results.order;
order = order * 2;

f = designfilt('bandstopiir','FilterOrder',order, ...
   'HalfPowerFrequency1',notch_freq-1,'HalfPowerFrequency2',notch_freq+1, ...
   'DesignMethod','butter','SampleRate',fs);
%fvtool(f)

for i = 1:size(values,2) 
    eeg = values(:,i);
    if sum(~isnan(eeg)) == 0
        continue
    end
    eeg(isnan(eeg)) = mean(eeg,'omitnan');
    eeg = filtfilt(f,eeg);   
    values(:,i) = eeg;
end


end

File: python/CNTtools/tools/notch_filter.py
from scipy.signal import iirnotch, sosfiltfilt, butter
import numpy as np
from beartype import beartype
from numbers import Number


@beartype
def notch_filter(
    data: np.ndarray, fs: Number, notch_freq: Number = 60, order: int = 4
) -> np.ndarray:
    """Apply notch filter to remove interference at a specified frequency.

    Args:
        values (np.ndarray): Matrix representing EEG data. Each column is a channel, and each row is a time point.
        fs (float): Sampling frequency of the EEG data.
        notch_freq (float, optional): Frequency to notch filter (default is 60 Hz).
        order (int, optional): Order of the notch filter (default is 4).

    Returns:
        np.ndarray: Matrix with notch-filtered EEG data.

    Example:
    values = np.random.randn(100, 5)  # Replace with your actual iEEG data
    fs = 250
    notch_filtered_values = notch_filter(values, fs)
    notch_filtered_values = notch_filter(values, fs, notch_freq=120, order=8)
    """
    # b, a = iirnotch(notch_freq, notch_freq / 2, fs=fs)
    sos = butter(
        order, [notch_freq - 1, notch_freq + 1], "bandstop", fs=fs, output="sos"
    )

    y = sosfiltfilt(sos, data, axis=0)

    return y
